lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,712 @@
+end
+
+--- Layout a grid layout.
+-- @param context The context in which we are drawn.
+-- @param width The available width.
+-- @param height The available height.
+function grid:layout(context, width, height)
+    local result = {}
+    local hspacing, vspacing = self._private.horizontal_spacing, self._private.vertical_spacing
+    local hpadding, vpadding = self._private.horizontal_padding, self._private.vertical_padding
+
+    -- Fit matrix cells
+    local rows_size, cols_size = get_matrix_sizes(self, context, width, height)
+    local total_expected_width, total_expected_height = 0, 0
+    for _,w in ipairs(cols_size) do
+        total_expected_width = total_expected_width + w
+    end
+    for _,h in ipairs(rows_size) do
+        total_expected_height = total_expected_height + h
+    end
+
+    -- Figure out the maximum size we can give out to sub-widgets
+    local sub_width = max_value(cols_size)
+    local sub_height = max_value(rows_size)
+    if self._private.horizontal_expand then
+        sub_width  = (width - 2*hpadding - (self._private.num_cols-1)*hspacing) / self._private.num_cols
+    end
+    if self._private.vertical_expand then
+        sub_height = (height - 2*vpadding - (self._private.num_rows-1)*vspacing) / self._private.num_rows
+    end
+
+    -- Calculate the position and size to place the widgets
+    local cumul_width, cumul_height = {}, {}
+    local cw, ch = hpadding, vpadding
+    for j = 1, #cols_size do
+        cumul_width[j] = cw
+        if self._private.horizontal_fixed_size then
+            cols_size[j] = math.max(self._private.min_cols_size, sub_width)
+        elseif self._private.horizontal_expand then
+            local hpercent = self._private.num_cols * sub_width * cols_size[j] / total_expected_width
+            cols_size[j] = math.max(self._private.min_cols_size, hpercent)
+        end
+        cw = cw + cols_size[j] + hspacing
+    end
+    cumul_width[#cols_size + 1] = cw
+    for i = 1, #rows_size do
+        cumul_height[i] = ch
+        if self._private.vertical_fixed_size then
+            rows_size[i] = math.max(self._private.min_rows_size, sub_height)
+        elseif self._private.vertical_expand then
+            local vpercent = self._private.num_rows * sub_height * rows_size[i] / total_expected_height
+            rows_size[i] = math.max(self._private.min_rows_size, vpercent)
+        end
+        ch = ch + rows_size[i] + vspacing
+    end
+    cumul_height[#rows_size + 1] = ch
+
+    -- Place widgets
+    local fill_space = true  -- should be fill_space property?
+    for _, v in pairs(self._private.widgets) do
+        local x, y, w, h
+        -- Round numbers to avoid decimals error, force to place tight widgets
+        -- and avoid redraw glitches
+        x = math.floor(cumul_width[v.col])
+        y = math.floor(cumul_height[v.row])
+        w = math.floor(cumul_width[v.col + v.col_span] - hspacing - x)
+        h = math.floor(cumul_height[v.row + v.row_span] - vspacing - y)
```

Comment:
What's a decimal error? Also, why are you `floor`ing the position, shouldn't that be `ceil`?